### PR TITLE
Auto failover3

### PR DESCRIPTION
**Pull request by mrm9084**
# Description

Adds Auto Failover to App Configuration. Looks up SRV records of configuration stores to find possible backup stores in case of provided stores are unreachable.

# All SDK Contribution checklist:
- [X] **The pull request does not introduce [breaking changes]**
- [X] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/autofailover/ReplicaLookUp.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+                if (configStore.getConnectionStrings().size() > 0) {
+                    providedEndpoints = configStore.getConnectionStrings().stream().map(connectionString -> {
+                        return (AppConfigurationReplicaClientsBuilder
+                            .getEndpointFromConnectionString(connectionString));
+                    }).toList();
+                } else if (configStore.getEndpoints().size() > 0) {
+                    providedEndpoints = configStore.getEndpoints();
+                } else {
+                    providedEndpoints = List.of(configStore.getEndpoint());
+                }
```
Comment (ivywei0125) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/autofailover/ReplicaLookUp.java:81]: If user want to use auto fail over, they need to define the endpoints in yaml file. I'm wonderring why do we just get all replacs and generate the `providedEndpoints` list by replicas? We could get all replicas by connection string.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/autofailover/ReplicaLookUp.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+                try {
+                    List<SRVRecord> srvRecords = findAutoFailoverEndpoints(mainEndpoint, providedEndpoints);
+
+                    srvRecords.sort((SRVRecord a, SRVRecord b) -> a.compareTo(b));
```
Comment (ivywei0125) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/autofailover/ReplicaLookUp.java:86]: Do we need to sort this `srvRecords`? Is there no priority order for the user-defined endpoints array in yaml file?
If we sort this array, then it doesn't follow the order that user defined.

Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/autofailover/ReplicaLookUp.java:86]: This came up in the design review. We plan on supporting the SRV priority and weight values. 

When a user defines multiple stores, we follow the list they provide, top down where the top option is the highest priority.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java [hunk 3]
```diff
@@ -134,13 +139,16 @@ List<AppConfigurationReplicaClient> buildClients(ConfigStore configStore) {
         int hasMultiConnectionString = configStore.getConnectionStrings().size() > 0 ? 1 : 0;
 
         if (hasSingleConnectionString + hasMultiEndpoints + hasMultiConnectionString > 1) {
-            throw new IllegalArgumentException("More than 1 Connection method was set for connecting to App Configuration.");
+            throw new IllegalArgumentException(
+                "More than 1 Connection method was set for connecting to App Configuration.");
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java:143]:
```suggestion
                "More than 1 connection method was set for connecting to App Configuration.");
```

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java [hunk 4]
```diff
@@ -134,13 +139,16 @@ List<AppConfigurationReplicaClient> buildClients(ConfigStore configStore) {
...
         }
 
-        boolean connectionStringIsPresent = configStore.getConnectionString() != null || configStore.getConnectionStrings().size() > 0;
+        boolean connectionStringIsPresent = configStore.getConnectionString() != null
+            || configStore.getConnectionStrings().size() > 0;
 
         if (credentialConfigured && connectionStringIsPresent) {
-            throw new IllegalArgumentException("More than 1 Connection method was set for connecting to App Configuration.");
+            throw new IllegalArgumentException(
+                "More than 1 Connection method was set for connecting to App Configuration.");
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java:151]:
```suggestion
                "More than 1 connection method was set for connecting to App Configuration.");
```

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java [hunk 5]
```diff
@@ -169,7 +177,9 @@ List<AppConfigurationReplicaClient> buildClients(ConfigStore configStore) {
                 if (!credentialConfigured) {
                     // System Assigned Identity. Needs to be checked last as all of the above should
                     // have an Endpoint.
-                    LOGGER.debug("Connecting to {} using Azure System Assigned Identity or Azure User Assigned Identity.", endpoint);
+                    LOGGER.debug(
+                        "Connecting to {} using Azure System Assigned Identity or Azure User Assigned Identity.",
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java:181]: Is it System/User Assigned **Managed** Identity?

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java [hunk 6]
```diff
@@ -182,8 +192,38 @@ List<AppConfigurationReplicaClient> buildClients(ConfigStore configStore) {
...
+            connectionString.setUri(failoverEndpoint);
+            ConfigurationClientBuilder builder = createBuilderInstance().connectionString(connectionString.toString());
+            return modifyAndBuildClient(builder, failoverEndpoint, 0);
+        } else {
+            ConfigurationClientBuilder builder = createBuilderInstance();
+            if (!credentialConfigured) {
+                // System Assigned Identity. Needs to be checked last as all of the above should
+                // have an Endpoint.
+                LOGGER.debug(
+                    "Connecting to {} using Azure System Assigned Identity or Azure User Assigned Identity.",
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java:213]: Same here

**Review by saragluna — APPROVED** [review]

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java (hunk 5 above)
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java]: System, I think this comment/log is from the old 2.x version. Updated.

File: sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java (hunk 6 above)
Comment (mrm9084) [reply sdk/spring/spring-cloud-azure-appconfiguration-config/src/main/java/com/azure/spring/cloud/appconfiguration/config/implementation/AppConfigurationReplicaClientsBuilder.java]: System, I think this comment/log is from the old 2.x version. Updated.

**Pull request merged**